lib/Dialect/HW/HWOps.cpp

```diff
@@ -2254,6 +2254,31 @@ OpFoldResult EnumConstantOp::fold(FoldAdaptor adaptor) {
   return getFieldAttr();
 }
 
+LogicalResult EnumConstantOp::verify() {
+  auto fieldAttr = getFieldAttr();
+  auto fieldType = fieldAttr.getType().getValue();
+  // This check ensures that we are using the exact same type, without looking
+  // through type aliases.
+  if (fieldType != getType())
+    emitOpError("return type ")
+        << getType() << " does not match attribute type " << fieldAttr;
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// EnumCmpOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult EnumCmpOp::verify() {
+  // Compare the canonical types.
+  auto lhsType = type_dyn_cast<EnumType>(getLhs().getType());
+  auto rhsType = type_dyn_cast<EnumType>(getRhs().getType());
```

Comment:
Not sure if need to check for null types, but since there's no `type_dyn_cast_or_null` shrug.  Hopefully something else catches that (ODS constraints, etc.) anyway.

Reply:
There are already type constraints in place that make sure the canonical type is a an enumeration, so we should be good there. On that note, I switched to `type_cast` here.